File: src/main/scala/li/cil/oc/common/item/traits/GPULike.scala
package li.cil.oc.common.item.traits

import li.cil.oc.Settings
import li.cil.oc.common.item.abstracts.SimpleItem
import li.cil.oc.util.PackedColor

import scala.jdk.CollectionConverters.SeqHasAsJava

trait GPULike extends SimpleItem {
  def gpuTier: Int

  override protected def tooltipData = {
    val (w, h) = (Settings.screenResolutionsByTier.get(gpuTier).getKey, Settings.screenResolutionsByTier.get(gpuTier).getValue)
    val depth = PackedColor.Depth.bits(Settings.screenDepthsByTier(gpuTier))
    Seq(w, h, depth,
      gpuTier match {
        case 0 => "1/1/4/2/2"
        case 1 => "2/4/8/4/4"
        case 2 => "4/8/16/8/8"
      })
  }.map(_.asInstanceOf[AnyRef]).asJava
}

File: src/main/scala/li/cil/oc/common/item/traits/GPULike.java
package li.cil.oc.common.item.traits;

import li.cil.oc.Settings;
import li.cil.oc.util.PackedColor;

import java.util.Arrays;
import java.util.List;

public interface GPULike {
    int gpuTier();

    default List<Object> gpuTooltipData() {
        int w = Settings.screenResolutionsByTier.get(gpuTier()).getKey();
        int h = Settings.screenResolutionsByTier.get(gpuTier()).getValue();
        int depth = PackedColor.Depth$.MODULE$.bits(Settings.screenDepthsByTier[gpuTier()]);

        switch (Math.max(0, Math.min(gpuTier(), 2))) {
            case 0: return Arrays.asList(w, h, depth, "1/1/4/2/2");
            case 1: return Arrays.asList(w, h, depth, "2/4/8/4/4");
            case 2: return Arrays.asList(w, h, depth, "4/8/16/8/8");
        }
        throw new RuntimeException("Unreachable code");
    }
}
